File: back-end/src/main/java/es/ull/project/application/usecase/infrastructureplan/UpdateInfrastructurePlanUseCase.java
package es.ull.project.application.usecase.infrastructureplan;

import es.ull.project.domain.entity.InfrastructurePlan;
import es.ull.project.domain.valueobject.cost.MaximumBudget;
import es.ull.project.domain.valueobject.policy.ServicePolicies;
import es.ull.project.domain.valueobject.time.PlanningPeriod;

import java.util.List;
import java.util.UUID;

/**
 * Use case for updating an infrastructure plan.
 */
public interface UpdateInfrastructurePlanUseCase {
    /**
     * Updates an existing infrastructure plan with the specified attributes.
     *
     * @param id                  the unique identifier of the infrastructure plan
     * @param newPeriod           the new planning period
     * @param newMaxBudget        the new maximum budget
     * @param newServicePolicies  the new service policies
     * @param selectedFacilityIds list of facility IDs to include in the plan (replaces existing)
     * @param serviceAssignmentIds list of service assignment IDs to include in the plan (replaces existing)
     * @return the updated infrastructure plan
     */
    InfrastructurePlan update(
        UUID id, 
        PlanningPeriod newPeriod, 
        MaximumBudget newMaxBudget, 
        ServicePolicies newServicePolicies,
        List<UUID> selectedFacilityIds,
        List<UUID> serviceAssignmentIds
    );
}

File: front-end/src/application/usecase/InfrastructurePlanManagement/update-infrastructure-plan-use-case.ts
// UpdateInfrastructurePlanUseCase.ts
// Use case contract for updating an infrastructure plan

import type { Either, DataError } from '@ull-tfg/ull-tfg-typescript';
import { InfrastructurePlan } from '../../../domain/entity/infrastructure-plan';
import { PlanningPeriod } from '../../../domain/valueobject/time/planning-period';
import { MaximumBudget } from '../../../domain/valueobject/cost/maximum-budget';
import { ServicePolicies } from '../../../domain/valueobject/policy/service-policies';

/**
 * Use case for updating an existing infrastructure plan in the system.
 * Input: planId and fields to update
 * Output: InfrastructurePlan (entity)
 */

// Command object for input data
export interface UpdateInfrastructurePlanCommand {
    planId: string;
    updatedFields: Partial<{
        period: PlanningPeriod;
        maxBudget: MaximumBudget;
        servicePolicies: ServicePolicies | null;
    }>;
}

// Result type for the use case
export type UpdateInfrastructurePlanResult = InfrastructurePlan;

// Use case contract
export interface UpdateInfrastructurePlanUseCase {
    /**
     * Handles the update of an existing infrastructure plan
     *
     * @param command Data needed to update the infrastructure plan
     * @returns Either a DataError or the updated InfrastructurePlan
     */
    execute(command: UpdateInfrastructurePlanCommand): Promise<Either<DataError, UpdateInfrastructurePlanResult>>;
}
